pliers/extractors/api/google.py

```diff
@@ -132,3 +138,233 @@ def _to_df(self, result):
+            extraction operation for completion.
+        discovery_file (str): path to discovery file containing Google
+            application credentials.
+        api_version (str): API version to use.
+        max_results (int): Max number of results per page.
+        num_retries (int): Number of times to retry query on failure.
+        rate_limit (int): The minimum number of seconds required between
+            transform calls on this Transformer.
+    '''
+
+    api_name = 'videointelligence'
+    _log_attributes = ('discovery_file', 'api_version', 'features', 'segments',
+                       'config', 'timeout', 'request_rate')
+
+    def __init__(self, features=['LABEL_DETECTION', 'SHOT_CHANGE_DETECTION',
+                                 'EXPLICIT_CONTENT_DETECTION'],
+                 segments=None, config=None, timeout=90, request_rate=5,
+                 discovery_file=None, api_version='v1', max_results=100,
+                 num_retries=3, rate_limit=None):
+        self.features = features
+        self.segments = segments
+        self.config = config
+        self.timeout = timeout
+        self.request_rate = request_rate
+        super(GoogleVideoIntelligenceAPIExtractor,
+              self).__init__(discovery_file=discovery_file,
+                             api_version=api_version,
+                             max_results=max_results,
+                             num_retries=num_retries,
+                             rate_limit=rate_limit)
+
+    def _query_api(self, request):
+        request_obj = self.service.videos().annotate(body=request)
+        return request_obj.execute(num_retries=self.num_retries)
+
+    def _query_operations(self, name):
+        request_obj = self.service.operations().get(name=name)
+        return request_obj.execute(num_retries=self.num_retries)
+
+    def _build_request(self, stim):
+        with stim.get_filename() as filename:
+            with open(filename, 'rb') as f:
+                vid_data = f.read()
+
+        context = self.config if self.config else {}
+        if self.segments:
+            context['segments'] = self.segments
+
+        request = {
+            'inputContent': base64.b64encode(vid_data).decode(),
+            'features': self.features,
+            'videoContext': context
+        }
+
+        return request
+
+    def _extract(self, stim):
+        op_request = self._build_request(stim)
+        operation = self._query_api(op_request)
+
+        msg = "Beginning video extraction with a timeout of %fs. Even for "\
+              "small videos, full extraction may take awhile." % self.timeout
+        logging.warning(msg)
+
+        operation_start = time.time()
+        response = self._query_operations(operation['name'])
+        while 'done' not in response and \
+              (time.time() - operation_start) < self.timeout:
+            response = self._query_operations(operation['name'])
+            time.sleep(self.request_rate)
```

Comment:
Polls the Google Video Intelligence Operation for completion every `self.request_rate` seconds